accepted/2020/platform-exclusion/platform-exclusion.md

```diff
@@ -0,0 +1,217 @@
+sandbox and thus can't freely create threads or open random files on disk.
+
+One could, in principle, apply the `[MinimumOSPlatform("...")]` attribute for
+all other OS platforms that do support them, but this has several downsides:
+
+1. The so marked APIs are now considered platform-specific, which means that
+   just calling the API from portable code will now always be flagged by the
+   analyzer.
+2. Every time .NET adds another operating system all call sides guarding calls
+   to these APIs now also have to check for this new OS.
+
+In this spec, we're proposing a mechanism by which we can mark APIs as
+unsupported.
+
+## Scenarios and User Experience
+
+* Building a Blazor Web Assembly app
+    - Using an API that is unsupported will be flagged by the analyzer.
+* Building a class library
+    - Building a regular class library (targeting `netstandard` or `net5.0`)
+      will not complain about APIs that are unsupported in Blazor.
+* Building a class library for use in Blazor Web Assembly
+    - Developer can modify the project file to explicitly indicate that they
+      want to see issues with APIs that are unsupported in Blazor.
+
+## Requirements
+
+### Goals
+
+* Be able to mark APIs that are unsupported on a particular OS
+* Allows platforms to be considered supported by default with the ability to add
+  or remove platforms from the project file and SDK.
+* Blazor Web Assembly is only considered supported when building a Blazor Web
+  Assembly app. For regular class libraries (`netstandard` or `net5.0`) the
+  default assumes that it's not supported and thus won't generate any warnings.
+  However, the developer must still be able to manually include Blazor Web
+  Assembly manually.
+* The annotations for unsupported APIs can express that an API was only
+  unsupported for a specific version.
+
+### Non-Goals
+
+## Design
+
+### Attribute
+
+```C#
+namespace System.Runtime.Versioning
+{
+    // Rename RemovedInOSPlatformAttribute to
+    // UnsupportedOSPlatformAttribute
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=true, Inherited=false)]
+    public sealed class UnsupportedOSPlatformAttribute : OSPlatformAttribute
+    {
+        public UnsupportedOSPlatformAttribute(string platformName);
+    }
+```
+
+The semantics of this attribute are as follows:
+
+* Marks an assembly, module, or API as unsupported on a given OS platform.
+* The `platformName` can contain no version number, a single version number, or
+  a range:
+    - `browser`. Marks the API as unsupported on all version of `browser`. This
```

Comment:
Note that versioning in `browser` is a bit awkward today. It is currently hard-coded to be `1.0.0.0`:

https://github.com/dotnet/runtime/blob/562763b5498bf53b68517b68221a89ad80ef6d16/src/libraries/System.Private.CoreLib/src/System/Environment.Browser.cs#L23

I'm not sure how we would version the "browser" OS.

cc @marek-safar @lewing @steveisok

Reply:
I don't think we would; it's an synthetic platform that basically means "the subset of .NET APIs that can run in the browser sandbox". The set and behaviors might be a moving target between .NET versions, but the semantic definition of the sandbox would remain fixed.

Reply:
If the browser sandbox evolves to support some of the concepts, would we just remove the attributes then?

Reply:
So are we saying that we think the "browser" OS is un-versionable?